### PR TITLE
[Op] Fix nll_loss

### PR DESCRIPTION
**Pull request by hgt312**
## Description ##
due to low precision of fp16, accumulation will produce strange value
change to use fp32 to accumulate, it may degrade the performance

```
In [10]: raf.nll_loss(rb, ra.to(dtype="float32"))
Out[10]:
[9.835546]
<NDArray [1] @ cpu, dtype=float32>

In [11]: raf.nll_loss(rb, ra)
Out[11]:
[4.]
<NDArray [1] @ cpu, dtype=float16>
```

P.S.: I also find that the grad of nll_loss have a bigger difference about 0.0007 when use fp16

## Checklist ##

- [x] PR's title starts with a category (e.g. [BUGFIX], [MODEL], [TUTORIAL], [FEATURE], [DOC], etc)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage
- [x] Code is well-documented

@yzhliu @comaniac

### CODE REVIEW

File: tests/python/op/tvm/test_tvm_loss.py [hunk 1]
```diff
@@ -83,10 +87,10 @@ def forward(self, y_true, y_pred):  # pylint: disable=no-self-use
     check(m_loss, t_loss)
     check(v_loss, t_loss)
     # backward
-    m_dy, t_dy = randn_torch((), device=device)
+    m_dy, t_dy = randn_torch((), device=device, dtype=dtype)
     t_loss.backward(t_dy)
     m_loss.backward(m_dy)
-    check(m_pred.grad, t_pred.grad)
+    check(m_pred.grad, t_pred.grad, atol=1e-3, rtol=1e-3)
```
Comment (comaniac) [inline tests/python/op/tvm/test_tvm_loss.py:93]: We should keep 1e-5 for fp32

File: python/raf/_tvm_op/loss.py [hunk 2]
```diff
@@ -83,13 +83,16 @@ def smooth_l1_loss_dtrue_compute(attr, inputs, output_type):  # pylint: disable=
 def nll_loss_compute(attrs, inputs, output_type):  # pylint: disable=unused-argument
     true, pred = inputs
     n, c = pred.shape
+    dtype = pred.dtype
+    if dtype == "float16":
+        pred = pred.astype("float32")
```
Comment (comaniac) [inline python/raf/_tvm_op/loss.py:88]: 1. Out of curiosity, what's the difference between `astype` and `cast` in TE compute?
2. If we now allow `nll_loss` to accept float16 inputs and implicitly turn them to float32 for accumulation, can we change the AMP rule accordingly (which forces `nll_loss` inputs to be float32 now)?

Comment (hgt312) [reply python/raf/_tvm_op/loss.py:88]: 1. they are the same
2. there are 3 main uses of `nll_loss`, 1) from_pytorch + nll_loss, with amp; 2) from_pytorch + nll_loss, pure fp16 model; 3) ratex lower. Change amp rule only covers 1).

Comment (comaniac) [reply python/raf/_tvm_op/loss.py:88]: For 2, I know that it only covers the first case you mentioned and that's exactly what I meant, although I guess it might not be beneficial because the softmax before it still runs with float32. Anyways, if there's no side effect then we could change that for better consistency.

Comment (hgt312) [reply python/raf/_tvm_op/loss.py:88]: `nll_loss` is already in "never cast"

**Comment by comaniac** [comment]
Thanks @hgt312
